WalletWasabi/Userfacing/AddressParser.cs

```diff
@@ -0,0 +1,96 @@
+using System.Collections.Generic;
+using NBitcoin;
+using WalletWasabi.Helpers;
+using WalletWasabi.Userfacing.Bip21;
+using WalletWasabi.Wallets.SilentPayment;
+using NBitcoinExtensions = WalletWasabi.Extensions.NBitcoinExtensions;
+
+namespace WalletWasabi.Userfacing;
+using AddressParsingResult = Result<Address, string>;
+
+public abstract record Address
+{
+	public record Bip21Uri(BitcoinAddress Address, decimal? Amount, string? Label, string? PayjoinEndpoint) : Address;
+	public record Bitcoin(BitcoinAddress Address) : Address;
+	public record SilentPayment(SilentPaymentAddress Address) : Address;
+
+	public string ToWif(Network network) =>
+		this switch
+		{
+			Bitcoin bitcoin => bitcoin.Address.ToString(),
+			SilentPayment sp => sp.Address.ToWip(network),
+			Bip21Uri bip21 => $"bitcoin:{bip21.Address}"
+			                  + (bip21.Amount is not null ? $"&amount={bip21.Amount}" : "")
+			                  + (bip21.Label is not null ? $"&label={bip21.Label}" : "")
+			                  + (bip21.PayjoinEndpoint is not null ? $"&pj={bip21.PayjoinEndpoint}" : "")
+		};
+}
+
+public static class AddressParser
+{
+	/// <summary>
+	/// Parses either a Bitcoin address or a BIP21 URI string.
+	/// </summary>
+	/// <seealso href="https://github.com/lightning/bolts/blob/master/11-payment-encoding.md"/>
+	public static AddressParsingResult Parse(string text, Network expectedNetwork)
+	{
+		text = text.Trim();
+
+		if (text == "")
+		{
+			return AddressParsingResult.Fail("Input length is invalid.");
+		}
+
+		// Too long URIs/Bitcoin address are unsupported.
+		if (text.Length > 1000)
+		{
+			return AddressParsingResult.Fail("Input is too long.");
+		}
+
+		// Parse a Bitcoin address (not BIP21 URI string)
+		if (!text.StartsWith($"{Bip21UriParser.UriScheme}:", StringComparison.OrdinalIgnoreCase))
+		{
+			Network networkGuess = expectedNetwork == Network.TestNet ? Network.Main : Network.TestNet;
+			if (NBitcoinExtensions.TryParseBitcoinAddressForNetwork(text, networkGuess, out _))
+			{
+				return AddressParsingResult.Fail($"Bitcoin address is valid for {networkGuess} and not for {expectedNetwork}.");
```

Comment:
it never returns this?
pasting mainnet addr while using testnet and vice versa just returns the "Input a valid BTC address or url"

Reply:
It never did before either. That check makes no sense. I will remove it.

Reply:
I was wrong. It should be shown while scanning a QR code.

Reply:
Even if it was not implemented, doing it is really easy with the "new design" using Options so I quickly added this feature which I believe is pretty great.